--- a/modules/bluetooth.mk
+++ b/modules/bluetooth.mk
@@ -70,6 +70,7 @@ include hardware/realtek/rtkbt/rtkbt.mk
 endif

 ifeq ($(strip $(BOARD_HAVE_BLUETOOTH_AIC)), true)
+ -include hardware/aic/aicbt/aicbt.mk
 -include hardware/aic/device-aic.mk
 PRODUCT_PACKAGES += libbt-vendor-aic
 endif


 
diff --git a/init.connectivity.rc b/init.connectivity.rc
index 4d8d102e..bfaffac4 100755
--- a/init.connectivity.rc
+++ b/init.connectivity.rc
@@ -42,6 +42,12 @@ on zygote-start
     chown bluetooth system /sys/class/rfkill/rfkill1/name
     chmod 0664 /sys/class/rfkill/rfkill1/state
 
+# for aic bt usb
+    chmod 0660 /dev/aic_btusb
+    chown bluetooth system /dev/aic_btusb
+    chmod 0660 /dev/aicbt_dev
+    chown bluetooth system /dev/aicbt_dev
+
 # for wifi
 
     mkdir /data/vendor/wifi 0771 wifi wifi
diff --git a/rootdir/init.insmod_charger.cfg b/rootdir/init.insmod_charger.cfg
index e361f664..0c18338e 100644
--- a/rootdir/init.insmod_charger.cfg
+++ b/rootdir/init.insmod_charger.cfg
@@ -1,2 +1,3 @@
 #insmod /rk30xxnand_ko.ko
+insmod /vendor/lib/modules/aic_btusb.ko
 
diff --git a/rootdir/ueventd.rockchip.rc b/rootdir/ueventd.rockchip.rc
index 0d5edd80..1b6fcfef 100644
--- a/rootdir/ueventd.rockchip.rc
+++ b/rootdir/ueventd.rockchip.rc
@@ -61,6 +61,8 @@ import /vendor/etc/ueventd.car.rc
 /dev/ttyS1                0660   bluetooth  system
 /dev/ttyS2                0660   bluetooth  system
 /dev/rtk_btusb            0660   bluetooth  system
+/dev/aic_btusb            0660   bluetooth  system
+/dev/aicbt_dev            0660   bluetooth  system
 
 #for hid audio
 /dev/hidraw0              0660   audio  audio
diff --git a/sepolicy/vendor/device.te b/sepolicy/vendor/device.te
index 2414381c..9841d024 100755
--- a/sepolicy/vendor/device.te
+++ b/sepolicy/vendor/device.te
@@ -5,6 +5,7 @@ type timer_device, dev_type;
 type vpu_device, dev_type;
 type hevc_device, dev_type;
 type rtkbt_device, dev_type;
+type aicbt_device, dev_type;
 type nvm_block_device, dev_type;
 # for rknand device
 type rknand_device, dev_type;
diff --git a/sepolicy/vendor/file_contexts b/sepolicy/vendor/file_contexts
index ac42e7f9..52acb74f 100644
--- a/sepolicy/vendor/file_contexts
+++ b/sepolicy/vendor/file_contexts
@@ -203,6 +203,7 @@
 /vendor/etc/init\.insmod\.cfg                                        u:object_r:init-insmod-sh_exec:s0
 
 /dev/rtkbt_dev       u:object_r:rtkbt_device:s0
+/dev/aicbt_dev       u:object_r:aicbt_device:s0
 
 #weaver hal
 /vendor/bin/hw/android\.hardware\.weaver@1\.0-service                u:object_r:weaver-hal-1-0_exec:s0
