diff --git a/modules/bluetooth.mk b/modules/bluetooth.mk
index 426e4af2..1395409c 100644
--- a/modules/bluetooth.mk
+++ b/modules/bluetooth.mk
@@ -30,8 +30,10 @@ PRODUCT_PACKAGES += \
 # Set supported Bluetooth profiles to enabled
 PRODUCT_PRODUCT_PROPERTIES += \
 	bluetooth.profile.asha.central.enabled?=true \
-	bluetooth.profile.a2dp.source.enabled?=true \
-	bluetooth.profile.avrcp.target.enabled?=true \
+	bluetooth.profile.a2dp.source.enabled?=false \
+	bluetooth.profile.avrcp.target.enabled?=false \
+	bluetooth.profile.a2dp.sink.enabled?=true \
+	bluetooth.profile.avrcp.controller.enabled?=true \
 	bluetooth.profile.bas.client.enabled?=true \
 	bluetooth.profile.gatt.enabled?=true \
 	bluetooth.profile.hfp.ag.enabled?=true \
@@ -91,10 +93,4 @@ PRODUCT_COPY_FILES += \
     $(TARGET_DEVICE_DIR)/bt_vendor.conf:/vendor/etc/bluetooth/bt_vendor.conf
